### PR TITLE
fix(poetry): normalize package name

### PR DESCRIPTION
**Pull request by dave-powell-kensho**
## Changes

(poetry) Normalize project names read from pyproject.toml in adherence with [PEP 503](https://peps.python.org/pep-0503/) and [PEP 621](https://peps.python.org/pep-0621/). poetry stores normalized project names in poetry.lock files, so without this normalization on the data read from the pyproject.toml files, there may be a misalignment leading to renovate not being able to update the lockfile.

For example, a valid entry in pyproject.toml may read as `prometheus_client`, but is represented in the lockfile as `prometheus-client`.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/poetry/extract.ts [hunk 1]
```diff
@@ -79,14 +81,14 @@ function extractFromSection(
       }
     }
     const dep: PackageDependency = {
-      depName,
+      depName: depNameNormalized,
```
Comment (rarkins) [inline lib/modules/manager/poetry/extract.ts:84]: Should the normalized name be used in lookups to e.g. pypi as well?

Comment (dave-powell-kensho) [reply lib/modules/manager/poetry/extract.ts:84]: According to PEP 503, yes:

> Below the root URL is another URL for each individual project contained within a repository. The format of this URL is /<project>/ where the <project> is replaced by the normalized name for that project

In practice, using the above example, navigating to https://pypi.org/simple/prometheus_client/ results in a redirect to https://pypi.org/simple/prometheus-client/

Now that I am looking at the documentation again, I see that the project name should also be lowercased:

> The name should be lowercased with all runs of the characters ., -, or _ replaced with a single - character.

Will update the PR to lowercase the normalized version as well.

Comment (dave-powell-kensho) [reply lib/modules/manager/poetry/extract.ts:84]: I don't have a strong opinion here. I went with changing the depName to the normalized one to achieve consistency on a name that is functionally correct for dependency management.

File: lib/modules/manager/poetry/extract.ts [hunk 2]
```diff
@@ -54,6 +54,8 @@ function extractFromSection(
       continue;
     }
 
+    let pep503NormalizeRegex = /[-_.]+/g;
```
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:57]:
```suggestion
    let pep503NormalizeRegex = regEx(/[-_.]+/g);
```
use our regEx wrapper

File: lib/modules/manager/poetry/extract.ts (hunk 1 above)
Comment (viceice) [reply lib/modules/manager/poetry/extract.ts:84]: i would prefer to leave depName as is an update packageName instead.

we already do that in other managers too. so it's consistent inside renovate

Comment (dave-powell-kensho) [reply lib/modules/manager/poetry/extract.ts:84]: Works for me, will have an update soon.

File: lib/modules/manager/poetry/extract.ts [hunk 3]
```diff
@@ -84,9 +87,10 @@ function extractFromSection(
       currentValue,
       managerData: { nestedVersion },
       datasource: PypiDatasource.id,
+      packageName,
```
Comment (rarkins) [inline lib/modules/manager/poetry/extract.ts:90]: Set packageName only if packageName !== depName

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — DISMISSED** [review]

**Review by viceice — DISMISSED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**